apps/gdalalg_raster_fillnodata.cpp

```diff
@@ -0,0 +1,180 @@
+    AddOverwriteArg(&m_overwrite);
+
+    AddBandArg(&m_band).SetDefault(m_band);
+
+    AddArg("max-distance", 'd',
+           _("The maximum distance (in pixels) that the algorithm will search "
+             "out for values to interpolate."),
+           &m_maxDistance)
+        .SetDefault(m_maxDistance)
+        .SetMetaVar("MAX_DISTANCE");
+
+    AddArg("smoothing-iterations", 's',
+           _("The number of 3x3 average filter smoothing iterations to run "
+             "after the interpolation to dampen artifacts. The default is zero "
+             "smoothing iterations."),
+           &m_smoothingIterations)
+        .SetDefault(m_smoothingIterations)
+        .SetMetaVar("SMOOTHING_ITERATIONS");
+
+    auto &mask{AddArg("mask", 0,
+                      _("Use the first band of the specified file as a "
+                        "validity mask (zero is invalid, non-zero is valid)."),
+                      &m_maskDataset)};
+
+    SetAutoCompleteFunctionForFilename(mask, GDAL_OF_RASTER);
+
+    mask.AddValidationAction(
+        [this, &mask]()
+        {
+            // Check if the mask dataset is a valid raster dataset descriptor
+            // Try to open the dataset as a raster
+            std::unique_ptr<GDALDataset> poDS(
+                GDALDataset::Open(mask.Get<std::string>().c_str(),
+                                  GDAL_OF_RASTER, nullptr, nullptr, nullptr));
+            return static_cast<bool>(poDS);
+        });
+
+    AddArg("strategy", 0,
+           _("By default, pixels are interpolated using an inverse distance "
+             "weighting (inv_dist). It is also possible to choose a nearest "
+             "neighbour (nearest) strategy."),
+           &m_strategy)
+        .SetDefault(m_strategy)
+        .SetChoices("inv_dist", "nearest");
+}
+
+/************************************************************************/
+/*                 GDALRasterFillNodataAlgorithm::RunImpl()             */
+/************************************************************************/
+
+bool GDALRasterFillNodataAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
+                                            void *pProgressData)
+{
+
+    VSIStatBufL sStat;
+    if (!m_overwrite && !m_outputDataset.GetName().empty() &&
+        (VSIStatL(m_outputDataset.GetName().c_str(), &sStat) == 0 ||
+         std::unique_ptr<GDALDataset>(
+             GDALDataset::Open(m_outputDataset.GetName().c_str()))))
+    {
+        ReportError(CE_Failure, CPLE_AppDefined,
+                    "File '%s' already exists. Specify the --overwrite "
+                    "option to overwrite it.",
+                    m_outputDataset.GetName().c_str());
+        return false;
+    }
+
+    CPLStringList aosOptions;
+    aosOptions.AddString("-of");
+    aosOptions.AddString(m_format.c_str());
+    aosOptions.AddString("-b");
+    aosOptions.AddString(CPLSPrintf("%d", m_band));
+
+    for (const auto &co : m_creationOptions)
```

Comment:
This is a suboptimal strategy in particular if outputing to a compressed dataset, because it will lead to the  band to being written twice leading to a larger file than needed.
I'd suggest to outputing to a temporary LZW  tiled compressed TIFF file (name generated with CPLGenerateTempFilename()), apply  GDALFillNodata() and finally translate to final file

Reply:
I'm not sure I understand what you mean, the steps here are:
1. copy the original raster band (only the required band) to a new dataset (because we don't want to alter the original)
2. apply fillnodata on the newly copied band

Are you suggesting instead to:
1. copy (and potentially change the format of) the original raster band to LZW tiled compressed TIFF (regardless of the user-required destination format which might be another one)
2. apply fillnodata on the  newly copied band
3. translate to the user required format

?

It seems to me potentially more expensive, but I admit I didn't look at the internals of fillnodata.

Reply:
> Are you suggesting instead to:

yes, and it is indeed more expansive, but otherwise offering creation option could be a lie to the user. We could potentially only do that only if compression is selected in creation options (and restrict that detection to GTiff output format), but that would be a complication.

Anyway, we can disregard that point for this PR, but then please create an enhancement ticket about the issue, and make sure to add a warning in the documentation.